code/modules/mob/living/death.dm

```diff
@@ -69,14 +69,17 @@
  * * gibbed - Was the mob gibbed?
 */
 /mob/living/proc/death(gibbed)
+	if(stat == DEAD)
+		return FALSE
+
 	set_stat(DEAD)
 	unset_machine()
 	timeofdeath = world.time
 	tod = station_time_timestamp()
-	var/turf/T = get_turf(src)
-	if(mind && mind.name && mind.active && !istype(T.loc, /area/centcom/ctf))
-		if(!isanimal_or_basicmob(src) || HAS_TRAIT(src, TRAIT_ALERT_GHOSTS_ON_DEATH))
-			deadchat_broadcast(" has died at <b>[get_area_name(T)]</b>.", "<b>[mind.name]</b>", follow_target = src, turf_target = T, message_type=DEADCHAT_DEATHRATTLE)
+	var/turf/death_turf = get_turf(src)
+	var/area/death_area = get_area(src)
+	if(mind && mind.name && mind.active && (!death_area || !(death_area.area_flags & NO_DEATH_MESSAGE)))
```

Comment:
Could you use `considered_alive` here in some way (or some other helper proc/variable), just feels like those first three vars run on and jumble a bit.